### PR TITLE
update coredns to v1.11.1

### PR DESCRIPTION
**Pull request by DockToFuture**
**How to categorize this PR?**

/area networking
/kind enhancement

**What this PR does / why we need it**:

Update CoreDNS to v1.11.1.
CoreDNS release notes: https://github.com/coredns/coredns/releases/tag/v1.11.0 and https://github.com/coredns/coredns/releases/tag/v1.11.1

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
For capabilities see: https://github.com/coredns/coredns/issues/6249
**Release note**:

```other operator
Update CoreDNS to v1.11.1.
```

### CODE REVIEW

File: imagevector/images.yaml [hunk 1]
```diff
@@ -296,7 +296,7 @@ images:
 - name: coredns
   sourceRepository: github.com/coredns/coredns
   repository: registry.k8s.io/coredns/coredns
-  tag: "v1.10.1"
+  tag: "v1.11.1"
```
Comment (ialidzhikov) [inline imagevector/images.yaml:317]: Can you also update the coredns version used in local setup?

Ref https://github.com/gardener/gardener/blob/328f202a312de139bb9ae80378c3641c254850c3/charts/gardener/provider-local/templates/coredns/deployment.yaml#L32

**Review by ialidzhikov — CHANGES_REQUESTED** [review]

File: imagevector/images.yaml [hunk 2]
```diff
@@ -293,10 +293,11 @@ images:
   sourceRepository: github.com/gardener/vpn2
   repository: eu.gcr.io/gardener-project/gardener/vpn-shoot-client
   tag: "0.20.0"
+# when updating coredns to v1.12 check if NET_BIND_SERVICE capability can be removed
```
Comment (ialidzhikov) [inline imagevector/images.yaml]:
```suggestion
# TODO(DockToFuture): When updating coredns to v1.12.x check if the NET_BIND_SERVICE capability can be removed.
```

**Review by ialidzhikov — CHANGES_REQUESTED** [review]

File: charts/gardener/provider-local/templates/coredns/deployment.yaml [hunk 3]
```diff
@@ -70,6 +70,8 @@ spec:
         securityContext:
           allowPrivilegeEscalation: false
           capabilities:
+            add:
+            - NET_BIND_SERVICE
```
Comment (ScheererJ) [inline charts/gardener/provider-local/templates/coredns/deployment.yaml:74]: Please add a comment that this should be removed with `coredns:v1.12`.

File: pkg/component/coredns/coredns.go [hunk 4]
```diff
@@ -406,6 +406,7 @@ import custom/*.server
 							SecurityContext: &corev1.SecurityContext{
 								AllowPrivilegeEscalation: pointer.Bool(false),
 								Capabilities: &corev1.Capabilities{
+									Add:  []corev1.Capability{"NET_BIND_SERVICE"},
```
Comment (ScheererJ) [inline pkg/component/coredns/coredns.go]: Please add a comment that this should be removed with `coredns:v1.12`.

**Review by ScheererJ — CHANGES_REQUESTED** [review]
Thanks a lot for the update. Please add `TODO` comments for the additional capabilities, though, as they should be dropped again later on.

File: charts/gardener/provider-local/templates/coredns/deployment.yaml (hunk 3 above)
Comment (DockToFuture) [reply charts/gardener/provider-local/templates/coredns/deployment.yaml:74]: I've already added a reminder to the images.yaml. I think it is not necessary to duplicate this to every occurrence in the code.

Comment (ScheererJ) [reply charts/gardener/provider-local/templates/coredns/deployment.yaml:74]: Personally, I prefer to tag all the locations that need to be changed, but I can also live with a single reminder in the images file.

**Comment by ScheererJ** [comment]
At least the `Unprivileged Shoot` test there seems to be a problem with `coredns`:

```
  {"level":"info","ts":"2023-12-13T08:03:10.667Z","logger":"shoot-test.test","msg":"Shoot not yet created","shoot":{"name":"e2e-unpriv","namespace":"garden-local"},"reason":"condition type SystemComponentsHealthy is not true yet, had message Deployment \"kube-system/coredns\" is unhealthy: condition \"Available\" has invalid status False (expected True) due to MinimumReplicasUnavailable: Deployment does not have minimum availability. with reason DeploymentUnhealthy"}
```

@DockToFuture Could you please take a look?

**Comment by DockToFuture** [comment]
> At least the `Unprivileged Shoot` test there seems to be a problem with `coredns`:
> 
> ```
>   {"level":"info","ts":"2023-12-13T08:03:10.667Z","logger":"shoot-test.test","msg":"Shoot not yet created","shoot":{"name":"e2e-unpriv","namespace":"garden-local"},"reason":"condition type SystemComponentsHealthy is not true yet, had message Deployment \"kube-system/coredns\" is unhealthy: condition \"Available\" has invalid status False (expected True) due to MinimumReplicasUnavailable: Deployment does not have minimum availability. with reason DeploymentUnhealthy"}
> ```
> 
> @DockToFuture Could you please take a look?

The tests are currently pinned to k8s verison 1.24.8, see https://github.com/gardener/gardener/blob/master/test/e2e/gardener/shoot/create_and_delete_unprivileged.go#L39. The documentation (https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted) states that the `NET_BIND_SERVICE` capability is only allowed with the restricted policy in kubernetes >= v1.25+ under linux. Thus it's currently not possible to fix the unprivileged tests for coredns v1.11.1 with the pinned kubernetes version.

**Comment by DockToFuture** [comment]
> > The tests are currently pinned to k8s verison 1.24.8, see https://github.com/gardener/gardener/blob/master/test/e2e/gardener/shoot/create_and_delete_unprivileged.go#L39. The documentation (https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted) states that the `NET_BIND_SERVICE` capability is only allowed with the restricted policy in kubernetes >= v1.25+ under linux. Thus it's currently not possible to fix the unprivileged tests for coredns v1.11.1 with the pinned kubernetes version.
> 
> There is open PR to drop K8s 1.24 support: #8989 (cc @shafeeqes). This PR removes the version pinning from the corresponding e2e test. Hence, I guess we can proceed with the coredns bump to v1.11.1 only when we drop support for K8s 1.24?

Yes

**Comment by shafeeqes** [comment]
> > The tests are currently pinned to k8s verison 1.24.8, see https://github.com/gardener/gardener/blob/master/test/e2e/gardener/shoot/create_and_delete_unprivileged.go#L39. The documentation (https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted) states that the `NET_BIND_SERVICE` capability is only allowed with the restricted policy in kubernetes >= v1.25+ under linux. Thus it's currently not possible to fix the unprivileged tests for coredns v1.11.1 with the pinned kubernetes version.
> 
> There is open PR to drop K8s 1.24 support: #8989 (cc @shafeeqes). This PR removes the version pinning from the corresponding e2e test. Hence, I guess we can proceed with the coredns bump to v1.11.1 only when we drop support for K8s 1.24?

Yes.

**Comment by shafeeqes** [comment]
Looks like #8989 won't be merged until v1.89 is released due to change in plans of the expiry of v1.24. [ref](https://github.com/gardener/gardener/pull/8989#issuecomment-1882734747)
What should we do then?

**Comment by ialidzhikov** [comment]
We could use `coredns@v1.10.1` for K8s <= 1.24 and `coredns@v1.11.1` (with the capability) for K8s > 1.24. WDYT?

**Comment by shafeeqes** [comment]
> We could use `coredns@v1.10.1` for K8s <= 1.24 and `coredns@v1.11.1` (with the capability) for K8s > 1.24. WDYT?

Yes. Makes sense.

**Comment by rfranzke** [comment]
@DockToFuture Can you please update this PR?

**Review by ScheererJ — APPROVED** [review]
/lgtm
/approve

**Pull request merged**